src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -54,19 +60,65 @@ class MyPyFieldSet(FieldSet):
     sources: PythonSources
 
 
+@dataclass(frozen=True)
+class MyPyPartition:
+    field_set_addresses: FrozenOrderedSet[Address]
+    closure: FrozenOrderedSet[Target]
+    interpreter_constraints: PexInterpreterConstraints
+    python_version_already_configured: bool
+
+
 class MyPyRequest(TypecheckRequest):
     field_set_type = MyPyFieldSet
 
 
-def generate_args(mypy: MyPy, *, file_list_path: str) -> Tuple[str, ...]:
+def generate_argv(
+    mypy: MyPy, *, file_list_path: str, python_version: Optional[str]
+) -> Tuple[str, ...]:
     args = []
     if mypy.config:
         args.append(f"--config-file={mypy.config}")
+    if python_version:
+        args.append(f"--python-version={python_version}")
     args.extend(mypy.args)
     args.append(f"@{file_list_path}")
     return tuple(args)
 
 
+def check_and_warn_if_python_version_configured(
+    *, config: Optional[FileContent], args: Tuple[str, ...]
+) -> bool:
+    ending = (
+        "Normally, Pants would automatically set this for you based on your code's interpreter "
+        "constraints (https://www.pantsbuild.org/docs/python-interpreter-compatibility).\n\n(This "
+        "allows Pants to partition your targets by their constraints, so that, for example, "
+        "you can run MyPy on Python 2-only code and Python 3-only code at the same time. This "
+        "feature may no longer work.)"
+    )
+    configured = False
+    if config and b"python_version" in config.content:
+        logger.warning(
+            f"You set `python_version` in {config.path}, which is used because of the "
+            f"`[mypy].config` option. {ending}"
+        )
+        configured = True
+    if "--py2" in args:
+        logger.warning(f"You set `--py2` in the `--mypy-args` option. {ending}")
+        configured = True
+    if any(arg.startswith("--python-version") for arg in args):
+        logger.warning(f"You set `--python-version` in the `--mypy-args` option. {ending}")
```

Comment:
Are we ever going to be printing this lengthy `ending` multiple times at all? Do we want to do that?

Reply:
No, we won't. Only one time. This is why we call the message in the rule that does the partitioning, rather than in the rule for each partition.

Reply:
What if `python_version` is in the config file, and `--py2` is in the args, and `--python-version X.Y.Z` is also in the CLI args?

Reply:
True. In that edge case, we'd log all 3 times. We can make this more clever - I don't think `--py2` and `--python-version` at the same time is likely, but it is feasible to have it in config and CLI at the same time because CLI overrides config. This is how `tox.ini` is set up.

Reply:
It doesn't seem too difficult to wait until the end and check:
```python
if configured:
  logger.warning(ending)
return configured
```